src/librustc_typeck/mem_categorization.rs

```diff
@@ -552,18 +572,23 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
     ) -> McResult<PlaceWithHirId<'tcx>> {
         debug!("cat_deref: base_place={:?}", base_place);
 
-        let base_ty = base_place.place.ty;
-        let deref_ty = match base_ty.builtin_deref(true) {
+        let base_ty = base_place.place.base_ty;
+        let base_curr_ty = base_place.place.ty();
+        let deref_ty = match base_curr_ty.builtin_deref(true) {
             Some(mt) => mt.ty,
             None => {
-                debug!("explicit deref of non-derefable type: {:?}", base_ty);
+                debug!("explicit deref of non-derefable type: {:?}", base_curr_ty);
                 return Err(());
             }
         };
         let mut projections = base_place.place.projections;
-        projections.push(Projection { kind: ProjectionKind::Deref(base_ty) });
+        projections.push(Projection {
+            kind: ProjectionKind::Deref,
+            before_ty: base_curr_ty,
+            after_ty: deref_ty,
+        });
 
-        let ret = PlaceWithHirId::new(node.hir_id(), deref_ty, base_place.place.base, projections);
+        let ret = PlaceWithHirId::new(node.hir_id(), base_ty, base_place.place.base, projections);
```

Comment:
nit: I'd just use `base_place.place.base_ty`, just to be consitent with `cat_projections`. nbd.

Reply:
Done